app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java

```diff
@@ -601,10 +611,22 @@ public void notifyDataSetChanged() {
     @Override
     public void retryUpload(Contribution contribution) {
         if (NetworkUtils.isInternetConnectionEstablished(getContext())) {
-            if (contribution.getState() == STATE_FAILED || contribution.getState() == STATE_PAUSED || contribution.getState()==Contribution.STATE_QUEUED_LIMITED_CONNECTION_MODE) {
-                contribution.setState(Contribution.STATE_QUEUED);
-                contributionsPresenter.saveContribution(contribution);
-                Timber.d("Restarting for %s", contribution.toString());
+            if (contribution.getState() == STATE_PAUSED || contribution.getState()==Contribution.STATE_QUEUED_LIMITED_CONNECTION_MODE) {
+                restartUpload(contribution);
+            } else if (contribution.getState() == STATE_FAILED) {
+                int retries = contribution.getRetries();
+                /* Limit the number of retries for a failed upload
+                   to handle cases like invalid filename as such uploads
+                   will never be successful */
+                if(retries < MAX_RETRIES) {
+                    contribution.setRetries(retries + 1);
+                    Timber.d("Retried %d times", retries + 1);
```

Comment:
Can it be changed to "Retried upload %d times" so it's clear what we're trying to retry?

Reply:
Indeed the debug log line could say that, and also say what file, because often several files are being retried at the same time.